rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -0,0 +1,294 @@
+package coop.rchain.rholang.interpreter
+
+import coop.rchain.rholang.interpreter.Env.DeBruijn
+
+object Substitute {
+
+  def subOrDec(term: Var)(implicit env: Env[Par]): Either[Var, Par] =
+    term match {
+      case BoundVar(level) =>
+        env.get(level) match {
+          case Some(par) => Right(rename(par, env.level))
+          case None =>
+            if (level - env.level < 0)
+              throw new Error(s"Illegal Index [Variable: $term, Level: ${level - env.level}]")
+            else Left(BoundVar(level - env.level))
+        }
+      case _ => throw new Error(s"Illegal Substitution [$term]")
+    }
+
+  def subOrDec(exp: EVar)(implicit env: Env[Par]): Either[EVar, Par] =
+    subOrDec(exp.v) match {
+      case Left(varue) => Left(EVar(varue))
+      case Right(par)  => Right(par)
+    }
+
+  def substitute(term: Channel)(implicit env: Env[Par]): Channel =
+    term match {
+      case Quote(p) => Quote(substitute(p))
+      case ChanVar(v) =>
+        subOrDec(v) match {
+          case Left(_v) => ChanVar(_v)
+          case Right(p) => Quote(p)
+        }
+    }
+
+  def substitute(term: Par)(implicit env: Env[Par]): Par = {
+
+    def subExp(expxs: List[Expr]): Par =
+      (Par() /: expxs) { (par, expr) =>
+        expr match {
+          case e @ EVar(_) =>
+            subOrDec(e) match {
+              case Left(_e)    => par :+ _e
+              case Right(_par) => par ++ _par
+            }
+          case e @ _ => par :+ substitute(e)
+        }
+      }
+
+    val sends   = for { s <- term.sends } yield substitute(s)
+    val recvs   = for { r <- term.receives } yield substitute(r)
+    val evals   = for { e <- term.evals } yield substitute(e)
+    val news    = for { n <- term.news } yield substitute(n)
+    val matches = for { m <- term.matches } yield substitute(m)
+
+    val freeCount: Int =
+      (0 /: sends)((i, s) => i + s.freeCount) +
+        (0 /: recvs)((i, r) => i + r.freeCount) +
+        (0 /: evals)((i, e) => i + e.freeCount) +
+        (0 /: news)((i, n) => i + n.freeCount) +
+        (0 /: matches)((i, m) => i + m.freeCount)
+
+    Par(
+      sends,
+      recvs,
+      evals,
+      news,
+      Nil,
+      matches,
+      term.id,
+      freeCount
+    ) ++ subExp(term.exprs)
+  }
+
+  def substitute(term: Send)(implicit env: Env[Par]): Send = {
+    val _data = term.data map { par =>
+      substitute(par)
+    }
+    val freeCount = term.chan.freeCount + (0 /: _data) { (i, par) =>
+      i + par.freeCount
+    }
+    Send(
+      substitute(term.chan),
+      _data,
+      term.persistent,
+      freeCount
+    )
+  }
+
+  def substitute(term: Receive)(implicit env: Env[Par]): Receive = {
+    val _binds = for { (xs, chan) <- term.binds } yield {
+      (xs, substitute(chan))
+    }
+    val _body = substitute(term.body)
+    val freeCount = _body.freeCount + (0 /: _binds) {
```

Comment:
Same comment here. FreeCount should be 0, and it's just an error if it isn't.

Reply:
See comment above. If processes with free variables can be subbed in, we need a recount.